packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart

```diff
@@ -27,53 +28,111 @@ import 'macos/native_assets.dart';
+      for (final Object? json in json['code_assets'] as List<Object?>? ?? const <Object?>[])
         CodeAsset.fromEncoded(EncodedAsset.fromJson(json! as Map<String, Object?>)),
     ];
-    return DartBuildResult(codeAssets, dependencies);
+    final List<DataAsset> dataAssets = <DataAsset>[
+      for (final Object? json in json['data_assets'] as List<Object?>? ?? const <Object?>[])
+        DataAsset.fromEncoded(EncodedAsset.fromJson(json! as Map<String, Object?>)),
+    ];
+    return DartHookResult(buildStart, buildEnd, codeAssets, dataAssets, dependencies);
   }
 
+  final DateTime buildStart;
+  final DateTime buildEnd;
   final List<CodeAsset> codeAssets;
+  final List<DataAsset> dataAssets;
   final List<Uri> dependencies;
 
   Map<String, Object?> toJson() => <String, Object?>{
+    'build_start': buildStart.toIso8601String(),
+    'build_end': buildEnd.toIso8601String(),
     'dependencies': <Object?>[for (final Uri dep in dependencies) dep.toString()],
     'code_assets': <Object?>[for (final CodeAsset code in codeAssets) code.encode().toJson()],
+    'data_assets': <Object?>[for (final DataAsset asset in dataAssets) asset.encode().toJson()],
   };
 
   /// The files that eventually should be bundled with the app.
   List<Uri> get filesToBeBundled => <Uri>[
     for (final CodeAsset code in codeAssets)
       if (code.linkMode is DynamicLoadingBundled) code.file!,
+    for (final DataAsset asset in dataAssets) asset.file,
   ];
+
+  /// Whether caller may need to re-run the dart build.
+  bool isUpToDate(FileSystem fileSystem) {
+    return !_wasAnyFileModifiedSince(fileSystem, buildStart, dependencies);
+  }
+
+  /// Whether the files produced by the build are up-to-date.
+  ///
+  /// NOTICE: The build itself may be up-to-date but the output may not be (as
+  /// the output may be existing on disc and not be produced by the build
+  /// itself - in which case we may not need to re-build if the file changes,
+  /// but we may need to make a new asset bundle with the modified file).
+  bool isOutputDirty(FileSystem fileSystem) {
+    return _wasAnyFileModifiedSince(fileSystem, buildEnd, filesToBeBundled);
+  }
+
+  static bool _wasAnyFileModifiedSince(FileSystem fileSystem, DateTime since, List<Uri> uris) {
+    for (final Uri uri in uris) {
+      final DateTime modified = fileSystem.statSync(uri.toFilePath()).modified;
+      if (modified.isAfter(since)) {
+        return true;
+      }
+    }
+    return false;
+  }
 }
 
-/// Invokes the build of all transitive Dart packages and prepares code assets
+/// Invokes the build of all transitive Dart package hooks and prepares assets
 /// to be included in the native build.
-Future<DartBuildResult> runFlutterSpecificDartBuild({
+Future<DartHookResult> runFlutterSpecificHooks({
   required Map<String, String> environmentDefines,
   required FlutterNativeAssetsBuildRunner buildRunner,
-  required TargetPlatform targetPlatform,
+  required TargetPlatform? targetPlatform,
```

Comment:
What does it mean for `targetPlatform` to be null? If it's necessary, we should document in which cases `null` is expected or what the behavior will be.

Reply:
Hm, I don't know why this was made nullable - removed the `?`.